WordPress/gutenberg
# Consolidate overall Global Styles mechanics in the server

This PR seeks to consolidate the overall server-side mechanism for Global Styles. Continues work from https://github.com/WordPress/gutenberg/pull/19883

1) It adds the user's Global Styles to the resolver, so it now works using core, theme, and user data. It stores the user's data using a Custom Post Type, so we can leverage existing editor mechanisms (undo/redo, draft/publish status for Global Styles, etc). This also presumes we want Global Styles to be tied to the theme so when theme changes we want to start with a new blank slate for Global Styles, which is implemented by using the theme's textdomain as part of the CPT's post name.

2) It exposes to the editor a couple of variables necessary for live-editing the styles: the ID of the entity that holds the user's styles + the base styles (core+theme). How this works will be demonstrated in subsequent PR. Note this is made to work only with the Beta Site Editor.

3) As per [past conversations](https://github.com/WordPress/gutenberg/pull/19883), it flattens the Global Styles tree stripping out the block variables. It looks like this:

```js
{
	'color': {
		'primary': '<primary color>',
		'background': '<background color>',
		'text': '<text color>',
	},
        'line-height': '<line height>'
}
```

which will be converted to this CSS rule (see [discussion about syntax](https://github.com/WordPress/gutenberg/pull/20047#discussion_r377547790)):

```js
:root {
	--wp--color--primary: '<primary color>';
	--wp--color--background: '<background color>';
	--wp--color--text: '<text color>';
        --wp--line-height: '<line height>';
}
```

Note that the names and variables we expose are still in flux, so take the current ones as only valid for demo purposes.

## How to test

- Install and enable one of the block-based themes (find one at [theme-experiments](https://github.com/WordPress/theme-experiments)). Instructions here assume Twenty Twenty Blocks.
- Check out this PR and `npm install && npm run build`.
- Create a **void** file named `experimental-theme.json` in your current theme's folder. For example, `wp-content/themes/twentytwenty-blocks/experimental-theme.json`.
- Got to Gutenberg > Experiments and enable "Full Site Editing".
- Go to Gutenberg > Site Editor.

**1) Check that core's Global Styles work:**

- Add a button and verify that its background color is `#52accc`.

**2) Check that theme's Global Styles override core's if present:**

- Add the following into the `experimental-theme.json` file created before.

```json
{
	"color": {
		"primary": "hotpink"
	}
}
```

- Reload the Site Editor.
- Add a button and verify that the button's color is now `hotpink`.

**3) Check that user's Global Styles override theme's if present:**

- Insert in the database a new post with the following data [1]:
  - post_type='wp_global_styles'
  - post_name='wp-global-styles-THEME-TEXTDOMAIN' (ex: 'wp-global-styles-twentytwenty-blocks')
  - post_content='{"color":{"primary":"blue"}}'
  - post_status='publish'
- Reload the Site Editor.
- Add a button and verify that the button's color is now `blue`.

---

[1] Make sure there isn't another post with the same post_name; if there is one, delete it first. Find below some ways you can add the required data for testing purposes. This is for the experimental Twenty Twenty Blocks theme, change to use your active theme's text-domain as required.

**wp-cli** -- you may need to provide other connection parameters depending on your setup:

```sh
wp post delete $(wp post list --post_type='wp_global_styles' --format=ids) --force
wp post create
    --post_type='wp_global_styles'
    --post_name='wp-global-styles-twentytwenty-blocks'
    --post_content='{"color":{"primary":"blue"}}'
    --post_status='publish'
```

**PHP** -- paste the following code within the `gutenberg_experimental_global_styles_enqueue_assets` when it comes to test user's Global Styles:

```php
	// Delete existing CPT, if any.
	$gs_cpt_list = wp_get_recent_posts( [
		'post_type' => 'wp_global_styles',
	] );
	foreach ( $gs_cpt_list as $gs_cpt ) {
		wp_delete_post( $gs_cpt['ID'] );
	}

	// Insert new one with required data.
	wp_insert_post( [
	    'ID'           => 0,
	    'post_content' => json_encode( [
	        'color'=> [
	            'primary' => 'blue',
	        ],
	    ] ),
	    'post_type'   => 'wp_global_styles',
	    'post_name'   => 'wp-global-styles-twentytwenty-blocks',
	    'post_status' => 'publish',
	] );
```
